eland/etl.py

```diff
@@ -205,33 +208,43 @@ def pandas_to_eland(
     else:
         es_client.indices.create(index=es_dest_index, body=mapping)
 
-    # Now add data
-    actions = []
-    n = 0
-    for row in pd_df.iterrows():
-        if es_dropna:
-            values = row[1].dropna().to_dict()
-        else:
-            values = row[1].to_dict()
-
-        if use_pandas_index_for_es_ids:
-            # Use index as _id
-            id = row[0]
-
-            # Use integer as id field for repeatable results
-            action = {"_index": es_dest_index, "_source": values, "_id": str(id)}
-        else:
-            action = {"_index": es_dest_index, "_source": values}
-
-        actions.append(action)
-
-        n = n + 1
-
-        if n % chunksize == 0:
-            bulk(client=es_client, actions=actions, refresh=es_refresh)
-            actions = []
+    def action_generator(
+        pd_df: pd.DataFrame,
+        es_dropna: bool,
+        use_pandas_index_for_es_ids: bool,
+        es_dest_index: str,
+    ) -> Generator[Dict[str, Any], None, None]:
+        for row in pd_df.iterrows():
+            if es_dropna:
+                values = row[1].dropna().to_dict()
+            else:
+                values = row[1].to_dict()
+
+            if use_pandas_index_for_es_ids:
+                # Use index as _id
+                id = row[0]
+
+                # Use integer as id field for repeatable results
+                action = {"_index": es_dest_index, "_source": values, "_id": str(id)}
+            else:
+                action = {"_index": es_dest_index, "_source": values}
+
+            yield action
+
+    # parallel_bulk is lazy generator so use deque to consume them immediately
+    deque(
+        parallel_bulk(
+            client=es_client,
+            actions=action_generator(
+                pd_df, es_dropna, use_pandas_index_for_es_ids, es_dest_index
+            ),
+            thread_count=thread_count,
+            chunk_size=chunksize / thread_count,
+            refresh=es_refresh,
```

Comment:
So instead of passing `refresh=` to `parallel_bulk()` which ends up waiting for a refresh to happen after every bulk call (can make things take forever!) we should add this to the end of the function:

```python
if es_refresh:
    es_client.indices.refresh(index=es_dest_index)
```